import os
import pandas as pd
from datetime import datetime
from modules import DatabaseConnection, Log, TelegramNotifier, DataProcessor, Monitoring
from utils import DataProcessingError, DatabaseConnectionError

def main():
    db = DatabaseConnection()
    notifier = TelegramNotifier()
    log = Log()

    try:
        notifier.send_message("üìä Iniciando o processo de carregamento de dados...")

        # Registrar o in√≠cio do processo
        data_inicio = datetime.now()

        df = pd.read_excel('data/sales_data_with_dates.xlsx')
        original_count = len(df)

        log.debug("‚úÖ Dados Carregados com sucesso", 'INFO', db.conn)

        log.debug("üåÄ Iniciando o processo de transforma√ß√£o de dados...", 'INFO', db.conn)

        processor = DataProcessor(df, db.conn)

        log.debug("üîç Iniciando o processo de valida√ß√£o dos dados...", 'INFO', db.conn)

        db.truncate_table('vendas.stage_sales_data') 

        df = processor.validate_data()

        log.debug("‚úÖ Valida√ß√µes de dados conclu√≠das.", 'INFO', db.conn)

        log.debug("üì• Iniciando o processo de inser√ß√£o dos dados...", 'INFO', db.conn)

        error_rows_df = processor.process_data()

        after_insert_count = processor.get_row_count()

        ignored_count = original_count - after_insert_count

        # Registrar o fim do processo
        data_fim = datetime.now()

        # Inserir os dados de monitoramento com o ID gerado, data_inicio e data_fim
        monitor = Monitoring(db.conn)
        monitor.insert_monitoring_data(original_count, after_insert_count, ignored_count, data_inicio, data_fim)

        processor.save_error_rows_to_csv(error_rows_df)

        # Verifica√ß√£o de erros
        if not error_rows_df.empty:  
            notifier.send_message(f"‚ö†Ô∏è Dados inseridos no banco com sucesso, mas {len(error_rows_df)} linhas apresentaram erros.")
        else:
            notifier.send_message("‚úÖ Dados inseridos no banco com sucesso.")

        log.debug('Script executado com sucesso.', 'INFO', db.conn)

    except DatabaseConnectionError as e:
        log.debug(f"Erro de conex√£o com o banco de dados: {e}", 'ERROR')
        notifier.send_message(f"‚ùå *Erro de conex√£o com o banco de dados:* {e}")
    except DataProcessingError as e:
        log.debug(f"Erro no processamento de dados: {e}", 'ERROR')
        notifier.send_message(f"‚ùå *Erro no processamento de dados:* {e}")
    except Exception as e:
        log.debug(f"Ocorreu um erro na execu√ß√£o do script: {e}", 'ERROR')
        notifier.send_message(f"‚ùå *Ocorreu um erro na execu√ß√£o do script:* {e}")
    finally:
        db.close()

if __name__ == '__main__':
    main()